File: osint_nexus/core/detectors/dns.py
from collections.abc import Mapping
from beartype import beartype

from osint_nexus.core.detectors.base import FingerprintStrategy
from osint_nexus.core.type_defs import JSONValue


class DnsFingerprintStrategy(FingerprintStrategy[Mapping[str, JSONValue], Mapping[str, JSONValue]]):
    """Strategy for DNS fingerprinting."""

    name: str = "dns_patterns"

    @beartype
    def extract(self, data: Mapping[str, JSONValue]) -> Mapping[str, JSONValue]:
        # Expecting data: {"resolver_ip": str, "query_types": list[str]}
        resolver = data.get("resolver_ip")
        query_types_raw = data.get("query_types")
        query_types: list[str] = query_types_raw if isinstance(query_types_raw, list) else []

        # Heuristic analysis: Resolver + Query pattern
        fingerprint: Mapping[str, JSONValue] = {
            "resolver": str(resolver) if isinstance(resolver, str) else None,
            "query_type_count": len(set(query_types)),
            "supports_dnssec": "DNSSEC" in query_types,
        }

        confidence = 0.6 if resolver is not None else 0.1

        return {
            "name": self.name,
            "data": fingerprint,
            "confidence": confidence,
        }

File: internal/engine/strategies/dns.go
package strategies

import (
	"context"

	"github.com/FJ-cyberzilla/osint-nexus/internal/types"
)

// DNSPayload implements types.FingerprintPayload for DNS fingerprinting input.
type DNSPayload struct {
	ResolverIP string   `json:"resolver_ip"`
	QueryTypes []string `json:"query_types"`
}

// PayloadType returns the payload type identifier.
func (p DNSPayload) PayloadType() string { return "dns_input" }

// DNSOutputPayload implements types.FingerprintPayload for DNS fingerprinting output.
type DNSOutputPayload struct {
	Resolver       string `json:"resolver"`
	QueryTypeCount int    `json:"query_type_count"`
	SupportsDNSSEC bool   `json:"supports_dnssec"`
}

// PayloadType returns the payload type identifier.
func (p DNSOutputPayload) PayloadType() string { return "dns_output" }

// DnsFingerprintStrategy implements FingerprintStrategy for DNS patterns.
type DnsFingerprintStrategy struct{}

// NewDnsFingerprintStrategy initializes a new DnsFingerprintStrategy.
func NewDnsFingerprintStrategy() *DnsFingerprintStrategy {
	return &DnsFingerprintStrategy{}
}

// Name returns the strategy name.
func (s *DnsFingerprintStrategy) Name() string {
	return "dns_patterns"
}

// Extract performs heuristic analysis on DNS data.
func (s *DnsFingerprintStrategy) Extract(ctx context.Context, data types.FingerprintData) (types.FingerprintResult, error) {
	payload, ok := data.Payload.(DNSPayload)
	if !ok {
		return types.FingerprintResult{
			Name: s.Name(),
			Data: types.FingerprintData{
				Type:    "dns_patterns",
				Payload: DNSOutputPayload{},
			},
			Confidence: 0.1,
		}, nil
	}

	queryTypes := make(map[string]struct{})
	supportsDNSSEC := false
	for _, s := range payload.QueryTypes {
		queryTypes[s] = struct{}{}
		if s == "DNSSEC" {
			supportsDNSSEC = true
		}
	}

	confidence := 0.1
	if payload.ResolverIP != "" {
		confidence = 0.6
	}

	return types.FingerprintResult{
		Name: s.Name(),
		Data: types.FingerprintData{
			Type: "dns_patterns",
			Payload: DNSOutputPayload{
				Resolver:       payload.ResolverIP,
				QueryTypeCount: len(queryTypes),
				SupportsDNSSEC: supportsDNSSEC,
			},
		},
		Confidence: confidence,
	}, nil
}
